### PR TITLE
Java: Add `XTRIM` command for streams

### PR DESCRIPTION
**Pull request by acarbonetto**
https://redis.io/docs/latest/commands/xtrim/

### CODE REVIEW

File: java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            return threshold.toString();
+        }
+    }
+
+    /**
+     * Converts options for {@link StreamBaseCommands#xtrim(String, TrimLimit)} into a String[].
+     *
+     * @return String[]
+     */
+    public static String[] createXtrimArgs(TrimLimit limit) {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java:146]: why not to make `public` instead?

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 2]
```diff
@@ -1649,11 +1648,17 @@ public void xadd(BaseClient client) {
...
         }
 
-        /**
-         * TODO add test to XTRIM on maxlen expect( await client.xtrim(key, { method: "maxlen",
-         * threshold: 1, exact: true, }), ).toEqual(1); expect(await client.customCommand(["XLEN",
-         * key])).toEqual(1);
-         */
+        // test xtrim to remove 1 element
+        assertEquals(1L, client.xtrim(key, new MaxLen(true, 1)).get());
+
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1653]: can you run xlen here?

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 3]
```diff
@@ -52,4 +53,24 @@ public interface StreamBaseCommands {
...
+     * // A nearly exact trimming of the stream to at least a length of 10, limited to 10 entries
+     * Long trimmed = client.xtrim("key", new StreamTrimOptions.MaxLen(false, 10L)).get();
+     * System.out.println("Number of trimmed entries from stream: " + trimmed);
+     *
+     * // An exact trimming of the stream by minimum id of "0-3", limit of 10 entries
+     * Long trimmed = client.xtrim("key", new StreamTrimOptions.MinId(true, "0-3", 10L)).get();
+     * System.out.println("Number of trimmed entries from stream: " + trimmed);
+     * }</pre>
+     */
+    CompletableFuture<Long> xtrim(String key, TrimLimit options);
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:75]: I think we should rename to `StreamTrimOptions`

Comment (acarbonetto) [reply java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:75]: Do you prefer that?

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 4]
```diff
@@ -52,4 +53,24 @@ public interface StreamBaseCommands {
...
+     * @example
+     *     <pre>{@code
+     * // A nearly exact trimming of the stream to at least a length of 10, limited to 10 entries
+     * Long trimmed = client.xtrim("key", new StreamTrimOptions.MaxLen(false, 10L)).get();
+     * System.out.println("Number of trimmed entries from stream: " + trimmed);
+     *
+     * // An exact trimming of the stream by minimum id of "0-3", limit of 10 entries
+     * Long trimmed = client.xtrim("key", new StreamTrimOptions.MinId(true, "0-3", 10L)).get();
+     * System.out.println("Number of trimmed entries from stream: " + trimmed);
+     * }</pre>
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:73]: Do we want our examples to print like this? I thought the convention was to use asserts and comments to indicate what is returned.

File: java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    public static final String TRIM_LIMIT_REDIS_API = "LIMIT";
+
+    public abstract static class TrimLimit {
+
+        /**
+         * If <code>true</code>, the stream will be trimmed exactly. Equivalent to <code>=</code> in the
+         * Redis API. Otherwise, the stream will be trimmed in a near-exact manner, which is more
+         * efficient, equivalent to <code>~</code> in the Redis API.
+         */
+        protected boolean exact;
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java:31]: Do we want this to be required? Redis defaults to exact trimming when unspecified, which doesn't seem to be an option here.

File: java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+         * Create a trim option to trim stream based on stream ID.
+         *
+         * @param exact Whether to match exactly on the threshold.
+         * @param threshold Comparison id.
+         * @param limit Max number of stream entries to be trimmed.
+         */
+        public MinId(boolean exact, @NonNull String threshold, long limit) {
+            this.threshold = threshold;
+            this.exact = exact;
+            this.limit = limit;
```
Comment (jonathanl-bq) [inline java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java:83]: Doesn't this mean you can specify "=" and limit at the same time? You should only be allowed to specify limit if using "~".

File: java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java (hunk 1 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java:146]: I don't understand. It is public.

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 2 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:1653]: xlen is not yet implemented.  Want me to add a xlen todo?

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java (hunk 4 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:73]: assert really only works when we know what the expected output would be.

File: java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java (hunk 5 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java:31]: Right.  It is optional.  Will fix.

File: java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java (hunk 6 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java:83]: You can.  If it's a request error on the Redis side, that's fine.  But the documentation doesn't indicate that there's a restriction here.

Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/commands/StreamTrimOptions.java:83]: Maybe I'll include an example in the IT for this...

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java (hunk 4 above)
Comment (jonathanl-bq) [reply java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:73]: Right, but usually the other PRs tend to just make up numbers. Either way, it's not a big deal. We can just let AWS decide if they like this or not.

**Review by jonathanl-bq — COMMENTED** [review]
Delete extra StreamTrimOptions.java and StreamAddOptions.java files now that they are consolidated into StreamOptions.java

**Review by jonathanl-bq — APPROVED** [review]

**Review by jonathanl-bq — APPROVED** [review]

File: java/client/src/main/java/glide/api/models/commands/StreamOptions.java [hunk 7]
```diff
@@ -8,42 +8,77 @@
...
+            }
 
-    /** If set, the add operation will also trim the older entries in the stream. */
-    private final StreamTrimOptions trim;
+            return optionArgs.toArray(new String[0]);
+        }
+    }
 
+    /**
+     * Optional arguments for {@link StreamBaseCommands#xtrim(String, StreamTrimOptions)}
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/StreamOptions.java:64]: SHould we mention xadd here too?

File: java/client/src/main/java/glide/api/models/commands/StreamOptions.java [hunk 8]
```diff
@@ -8,42 +8,77 @@
...
+        public static final String TRIM_NOT_EXACT_REDIS_API = "~";
+        public static final String TRIM_LIMIT_REDIS_API = "LIMIT";
+
         /**
          * If <code>true</code>, the stream will be trimmed exactly. Equivalent to <code>=</code> in the
          * Redis API. Otherwise, the stream will be trimmed in a near-exact manner, which is more
          * efficient, equivalent to <code>~</code> in the Redis API.
          */
-        protected boolean exact;
+        protected Boolean exact;
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/StreamOptions.java:81]: Maybe we should use 3 state enum here instead

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 9]
```diff
@@ -2676,6 +2676,109 @@ public void xadd_with_options_returns_success(Pair<StreamAddOptions, String[]> o
         assertEquals(returnId, payload);
     }
 
+    @Test
+    @SneakyThrows
+    public void xtrim_with_exact_MinId() {
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/RedisClientTest.java:2681]: Leave here test for `xtrim` only and move extreme options test to another place (another file)
Update `xadd` tests too

File: java/client/src/test/java/glide/api/models/TransactionTests.java [hunk 10]
```diff
@@ -393,6 +397,12 @@ public void transaction_builds_protobuf_request(BaseTransaction<?> transaction)
         transaction.xadd("key", Map.of("field1", "foo1"), StreamAddOptions.builder().id("id").build());
         results.add(Pair.of(XAdd, buildArgs("key", "id", "field1", "foo1")));
 
+        transaction.xadd("key", Map.of("field1", "foo1"), StreamAddOptions.builder().id("id").build());
+        results.add(Pair.of(XAdd, buildArgs("key", "id", "field1", "foo1")));
+
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/models/TransactionTests.java:402]:
```suggestion
```
repeated

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 11]
```diff
@@ -1610,9 +1613,7 @@ public void xadd(BaseClient client) {
                         .xadd(
                                 key,
                                 Map.of(field1, "foo3", field2, "bar3"),
-                                StreamAddOptions.builder()
-                                        .trim(new StreamAddOptions.MaxLen(Boolean.TRUE, 2L))
-                                        .build())
+                                StreamAddOptions.builder().trim(new MaxLen(Boolean.TRUE, 2L)).build())
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1616]: This should work too
```suggestion
                                StreamAddOptions.builder().trim(new MaxLen(true, 2L)).build())
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:1653]: yes or custom command + todo

File: java/client/src/test/java/glide/api/RedisClientTest.java (hunk 9 above)
Comment (acarbonetto) [reply java/client/src/test/java/glide/api/RedisClientTest.java:2579]: moved to StreamCommandsTest.java

File: java/client/src/main/java/glide/api/models/commands/StreamOptions.java (hunk 8 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/commands/StreamOptions.java:81]: yes missing (not specified), true, false

File: java/client/src/test/java/glide/api/RedisClientTest.java (hunk 9 above)
Comment (Yury-Fridlyand) [reply java/client/src/test/java/glide/api/RedisClientTest.java:2577]: My idea was to keep there test for `xadd` (and others) with minimal arguments or empty options AND test how non-empty options converted to command args in another test.
You moved entire test to another file, but you still combine 2 independent tests into one.

File: java/client/src/test/java/glide/api/StreamCommandsTest.java [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+                                    "*"
+                                })));
+    }
+
+    @SneakyThrows
+    @ParameterizedTest
+    @MethodSource("getStreamAddOptions")
+    public void xadd_with_options_returns_success(Pair<StreamAddOptions, String[]> optionAndArgs) {
+        // setup
+        String key = "testKey";
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/StreamCommandsTest.java:122]: Keep here only
```java
assertEquals(expectedArgs, options.toArgs());
```
And move all rest part (with mock, future and protobuf) back.

File: java/client/src/test/java/glide/api/StreamCommandsTest.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    public void setUp() {
+        connectionManager = mock(ConnectionManager.class);
+        commandManager = mock(CommandManager.class);
+        service = new RedisClient(connectionManager, commandManager);
+    }
+
+    private static List<Arguments> getStreamAddOptions() {
+        return List.of(
+                Arguments.of(
+                        Pair.of(
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/StreamCommandsTest.java:55]: Don't use pairs, junit will convert 2 objects in `Arguments` into 2 arguments of a parametrized test.

File: java/client/src/test/java/glide/api/StreamCommandsTest.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+                                    TRIM_NOT_EXACT_REDIS_API,
+                                    Long.toString(5L),
+                                    "*"
+                                })));
+    }
+
+    @SneakyThrows
+    @ParameterizedTest
+    @MethodSource("getStreamAddOptions")
+    public void xadd_with_options_returns_success(Pair<StreamAddOptions, String[]> optionAndArgs) {
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/StreamCommandsTest.java:120]:
```suggestion
    public void xadd_with_options_returns_success(StreamAddOptions options, String[] expectedArgs) {
```

File: java/client/src/test/java/glide/api/StreamCommandsTest.java [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+                                    NO_MAKE_STREAM_REDIS_API,
+                                    TRIM_MINID_REDIS_API,
+                                    TRIM_NOT_EXACT_REDIS_API,
+                                    Long.toString(5L),
+                                    "*"
+                                })));
+    }
+
+    @SneakyThrows
+    @ParameterizedTest
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/StreamCommandsTest.java:118]: nit: add third argument to the function - test name - and use it there
```suggestion
    @ParameterizedTest(name = "{0}")
```
(the index is zero based)

File: java/client/src/test/java/glide/api/StreamCommandsTest.java (hunk 14 above)
Comment (Yury-Fridlyand) [reply java/client/src/test/java/glide/api/StreamCommandsTest.java:120]: with test name
```suggestion
    public void xadd_with_options_returns_success(String testName, StreamAddOptions options, String[] expectedArgs) {
```

File: java/client/src/test/java/glide/api/StreamCommandsTest.java [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+                                new MinId("0-3", 10L),
+                                new String[] {
+                                    TRIM_MINID_REDIS_API, TRIM_NOT_EXACT_REDIS_API, "0-3", TRIM_LIMIT_REDIS_API, "10"
+                                })));
+    }
+
+    @SneakyThrows
+    @ParameterizedTest
+    @MethodSource("getStreamTrimOptions")
+    public void xtrim_with_options_returns_success(Pair<StreamTrimOptions, String[]> optionAndArgs) {
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/StreamCommandsTest.java:183]: same here

File: java/client/src/main/java/glide/api/models/commands/StreamOptions.java (hunk 8 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/commands/StreamOptions.java:81]: I think what we have is fine...

File: java/client/src/test/java/glide/api/StreamCommandsTest.java (hunk 13 above)
Comment (acarbonetto) [reply java/client/src/test/java/glide/api/StreamCommandsTest.java:55]: oh right

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 17]
```diff
@@ -2648,16 +2608,34 @@ private static List<Arguments> getStreamAddOptions() {
...
+
+    @SneakyThrows
+    @Test
+    public void xadd_with_nomakestream_maxlen_options_returns_success() {
         // setup
         String key = "testKey";
         Map<String, String> fieldValues = new LinkedHashMap<>();
         fieldValues.put("testField1", "testValue1");
         fieldValues.put("testField2", "testValue2");
+        StreamAddOptions options =
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/RedisClientTest.java:2626]: you can use empty options here to simplify the test

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 18]
```diff
@@ -2648,16 +2608,34 @@ private static List<Arguments> getStreamAddOptions() {
     }
 
     @SneakyThrows
-    @ParameterizedTest
+    @ParameterizedTest(name = "{0}")
     @MethodSource("getStreamAddOptions")
-    public void xadd_with_options_returns_success(Pair<StreamAddOptions, String[]> optionAndArgs) {
+    public void xadd_with_options_to_arguments(
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/RedisClientTest.java:2613]: excellent
optionally, we can move this to a separate file (we test options here, not client)

**Review by Yury-Fridlyand — APPROVED** [review]

File: java/client/src/test/java/glide/api/RedisClientTest.java (hunk 18 above)
Comment (acarbonetto) [reply java/client/src/test/java/glide/api/RedisClientTest.java:2613]: I was thinking that too.  It this becomes a pattern, I will move the tests.

File: java/client/src/test/java/glide/api/RedisClientTest.java (hunk 17 above)
Comment (acarbonetto) [reply java/client/src/test/java/glide/api/RedisClientTest.java:2626]: 🤷

**Pull request merged**